File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();

        int i = 0,j = 0,ans = 0;

        unordered_map<char,int>mp;

        while(j < n){
            while(mp[s[j]] != 0){
                mp[s[i++]]--;
            }
            mp[s[j]]++;
            ans = max(ans,j - i +1);

            j++;
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLength = 0;
        Map<Character, Integer> charMap = new HashMap<>();
        int left = 0;
        
        for (int i = 0; i < n; i++) {
            char currentChar = s.charAt(i);
            
            if (charMap.containsKey(currentChar)) {
            
                left = Math.max(left, charMap.get(currentChar) + 1);
            }
            
            
            charMap.put(currentChar, i);
            
            
            maxLength = Math.max(maxLength, i - left + 1);
        }
        
        return maxLength;
    }
}